File: clients/ios/Lextures/Core/LMS/NotificationLogic.swift
import Foundation

// MARK: - Categories & filters

enum NotificationCategory: String, CaseIterable, Identifiable, Hashable {
    case grades
    case assignments
    case discussions
    case announcements
    case messages
    case reminders
    case account
    case courses
    case other

    var id: String { rawValue }
}

enum NotificationFilter: String, CaseIterable, Identifiable, Hashable {
    case all
    case unread
    case grades
    case assignments
    case discussions
    case announcements
    case messages
    case reminders

    var id: String { rawValue }

    var labelKey: String {
        switch self {
        case .all: return "mobile.notifications.filter.all"
        case .unread: return "mobile.notifications.filter.unread"
        case .grades: return "mobile.notifications.category.grades"
        case .assignments: return "mobile.notifications.category.assignments"
        case .discussions: return "mobile.notifications.category.discussions"
        case .announcements: return "mobile.notifications.category.announcements"
        case .messages: return "mobile.notifications.category.messages"
        case .reminders: return "mobile.notifications.category.reminders"
        }
    }

    var localizedLabel: String {
        switch self {
        case .all: return L.text("mobile.notifications.filter.all")
        case .unread: return L.text("mobile.notifications.filter.unread")
        case .grades: return L.text("mobile.notifications.category.grades")
        case .assignments: return L.text("mobile.notifications.category.assignments")
        case .discussions: return L.text("mobile.notifications.category.discussions")
        case .announcements: return L.text("mobile.notifications.category.announcements")
        case .messages: return L.text("mobile.notifications.category.messages")
        case .reminders: return L.text("mobile.notifications.category.reminders")
        }
    }
}

enum NotificationLogic {
    static func category(for eventType: String) -> NotificationCategory {
        switch eventType {
        case "grade_posted":
            return .grades
        case "assignment_created", "assignment_due_reminder", "submission_received",
             "incomplete_granted", "incomplete_reminder":
            return .assignments
        case "discussion_reply":
            return .discussions
        case "course_announcement", "meeting_reminder", "conference_confirmed",
             "conference_reminder", "coaching_tip_weekly":
            return .announcements
        case "inbox_message":
            return .messages
        case "study_reminder_daily", "study_reminder_streak_at_risk", "study_reminder_weekly_summary":
            return .reminders
        case "password_reset", "welcome_invite", "payment_failed", "ceu_awarded", "certificate_issued":
            return .account
        case "canvas_course_imported", "course_copy_imported", "course_copy_import_failed":
            return .courses
        default:
            return .other
        }
    }

    static func matchesFilter(_ notification: AppNotification, filter: NotificationFilter) -> Bool {
        switch filter {
        case .all:
            return true
        case .unread:
            return !notification.isRead
        case .grades:
            return category(for: notification.eventType) == .grades
        case .assignments:
            return category(for: notification.eventType) == .assignments
        case .discussions:
            return category(for: notification.eventType) == .discussions
        case .announcements:
            return category(for: notification.eventType) == .announcements
        case .messages:
            return category(for: notification.eventType) == .messages
        case .reminders:
            return category(for: notification.eventType) == .reminders
        }
    }

    static func filter(_ notifications: [AppNotification], by filter: NotificationFilter) -> [AppNotification] {
        notifications.filter { matchesFilter($0, filter: filter) }
    }

    static func eventLabelKey(for eventType: String) -> String {
        "mobile.notifications.event.\(eventType)"
    }

    static func eventLabel(for eventType: String) -> String {
        let key = eventLabelKey(for: eventType)
        let localized = String(
            localized: String.LocalizationValue(stringLiteral: key),
            locale: LocalePreferences.effectiveLocaleValue()
        )
        return localized == key ? eventType.replacingOccurrences(of: "_", with: " ") : localized
    }

    static func categoryLabel(for category: NotificationCategory) -> String {
        switch category {
        case .grades: return L.text("mobile.notifications.category.grades")
        case .assignments: return L.text("mobile.notifications.category.assignments")
        case .discussions: return L.text("mobile.notifications.category.discussions")
        case .announcements: return L.text("mobile.notifications.category.announcements")
        case .messages: return L.text("mobile.notifications.category.messages")
        case .reminders: return L.text("mobile.notifications.category.reminders")
        case .account: return L.text("mobile.notifications.category.account")
        case .courses: return L.text("mobile.notifications.category.courses")
        case .other: return L.text("mobile.notifications.category.other")
        }
    }

    static func groupedPreferences(_ preferences: [NotificationPreference]) -> [(NotificationCategory, [NotificationPreference])] {
        let grouped = Dictionary(grouping: preferences) { category(for: $0.eventType) }
        return NotificationCategory.allCases.compactMap { category in
            guard let rows = grouped[category], !rows.isEmpty else { return nil }
            return (category, rows.sorted { $0.eventType < $1.eventType })
        }
    }

    static func isPushEnabled(eventType: String, preferences: [NotificationPreference]) -> Bool {
        preferences.first(where: { $0.eventType == eventType })?.pushEnabled ?? true
    }

    static func isEmailEnabled(eventType: String, preferences: [NotificationPreference]) -> Bool {
        preferences.first(where: { $0.eventType == eventType })?.emailEnabled ?? true
    }
}

// MARK: - Preferences cache (push gating + offline reads)

enum NotificationPreferencesCache {
    private static let storageKey = "notification_preferences_cache"

    static func save(_ preferences: [NotificationPreference], ownerKey: String) {
        guard let data = try? JSONEncoder().encode(preferences) else { return }
        UserDefaults.standard.set(data, forKey: key(for: ownerKey))
    }

    static func load(ownerKey: String) -> [NotificationPreference] {
        guard let data = UserDefaults.standard.data(forKey: key(for: ownerKey)),
              let decoded = try? JSONDecoder().decode([NotificationPreference].self, from: data) else {
            return []
        }
        return decoded
    }

    static func clear(ownerKey: String) {
        UserDefaults.standard.removeObject(forKey: key(for: ownerKey))
    }

    private static func key(for ownerKey: String) -> String {
        "\(storageKey).\(ownerKey)"
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/NotificationLogic.kt
package com.lextures.android.core.lms

enum class NotificationCategory {
    Grades,
    Assignments,
    Discussions,
    Announcements,
    Messages,
    Reminders,
    Account,
    Courses,
    Other,
}

enum class NotificationFilter(val id: String, val labelKey: String) {
    All("all", "mobile.notifications.filter.all"),
    Unread("unread", "mobile.notifications.filter.unread"),
    Grades("grades", "mobile.notifications.category.grades"),
    Assignments("assignments", "mobile.notifications.category.assignments"),
    Discussions("discussions", "mobile.notifications.category.discussions"),
    Announcements("announcements", "mobile.notifications.category.announcements"),
    Messages("messages", "mobile.notifications.category.messages"),
    Reminders("reminders", "mobile.notifications.category.reminders"),
}

object NotificationLogic {
    fun category(eventType: String): NotificationCategory = when (eventType) {
        "grade_posted" -> NotificationCategory.Grades
        "assignment_created", "assignment_due_reminder", "submission_received",
        "incomplete_granted", "incomplete_reminder" -> NotificationCategory.Assignments
        "discussion_reply" -> NotificationCategory.Discussions
        "course_announcement", "meeting_reminder", "conference_confirmed",
        "conference_reminder", "coaching_tip_weekly" -> NotificationCategory.Announcements
        "inbox_message" -> NotificationCategory.Messages
        "study_reminder_daily", "study_reminder_streak_at_risk", "study_reminder_weekly_summary" ->
            NotificationCategory.Reminders
        "password_reset", "welcome_invite", "payment_failed", "ceu_awarded", "certificate_issued" ->
            NotificationCategory.Account
        "canvas_course_imported", "course_copy_imported", "course_copy_import_failed" ->
            NotificationCategory.Courses
        else -> NotificationCategory.Other
    }

    fun matchesFilter(notification: AppNotification, filter: NotificationFilter): Boolean = when (filter) {
        NotificationFilter.All -> true
        NotificationFilter.Unread -> !notification.isRead
        NotificationFilter.Grades -> category(notification.eventType) == NotificationCategory.Grades
        NotificationFilter.Assignments -> category(notification.eventType) == NotificationCategory.Assignments
        NotificationFilter.Discussions -> category(notification.eventType) == NotificationCategory.Discussions
        NotificationFilter.Announcements -> category(notification.eventType) == NotificationCategory.Announcements
        NotificationFilter.Messages -> category(notification.eventType) == NotificationCategory.Messages
        NotificationFilter.Reminders -> category(notification.eventType) == NotificationCategory.Reminders
    }

    fun filter(notifications: List<AppNotification>, filter: NotificationFilter): List<AppNotification> =
        notifications.filter { matchesFilter(it, filter) }

    fun eventLabelKey(eventType: String): String = "mobile.notifications.event.$eventType"

    fun groupedPreferences(preferences: List<NotificationPreference>): List<Pair<NotificationCategory, List<NotificationPreference>>> {
        val grouped = preferences.groupBy { category(it.eventType) }
        return NotificationCategory.entries.mapNotNull { category ->
            val rows = grouped[category]?.sortedBy { it.eventType }.orEmpty()
            if (rows.isEmpty()) null else category to rows
        }
    }

    fun isPushEnabled(eventType: String, preferences: List<NotificationPreference>): Boolean =
        preferences.firstOrNull { it.eventType == eventType }?.pushEnabled ?: true

    fun isEmailEnabled(eventType: String, preferences: List<NotificationPreference>): Boolean =
        preferences.firstOrNull { it.eventType == eventType }?.emailEnabled ?: true
}
